paymentsheet/src/test/java/com/stripe/android/link/ui/paymentmethod/PaymentMethodViewModelTest.kt

```
@@ -69,49 +78,174 @@ class PaymentMethodViewModelTest {
+        assertThat(viewModel.state.value.errorMessage).isEqualTo(R.string.stripe_something_went_wrong.resolvableString)
+        assertThat(logger.errorLogs).containsExactly(
+            "PaymentMethodViewModel: Failed to create card payment details" to error
+        )
+    }
+
+    @Test
+    fun `onPayClicked handles confirmation failure`() = runTest {
+        val linkConfirmationHandler = FakeLinkConfirmationHandler()
+
+        linkConfirmationHandler.confirmResult = LinkConfirmationResult.Failed("Payment failed".resolvableString)
+
+        val viewModel = createViewModel(linkConfirmationHandler = linkConfirmationHandler)
+
+        viewModel.formValuesChanged(
+            formValues = FormFieldValues(userRequestedReuse = PaymentSelection.CustomerRequestedSave.NoRequest)
+        )
+
+        viewModel.onPayClicked()
+
+        assertThat(linkConfirmationHandler.calls).hasSize(1)
+        assertThat(viewModel.state.value.primaryButtonState).isEqualTo(PrimaryButtonState.Enabled)
+        assertThat(viewModel.state.value.errorMessage).isEqualTo("Payment failed".resolvableString)
+    }
+
+    @Test
+    fun `onPayClicked handles cancellation`() = runTest {
+        val linkConfirmationHandler = FakeLinkConfirmationHandler()
+        linkConfirmationHandler.confirmResult = LinkConfirmationResult.Canceled
+
+        val viewModel = createViewModel(linkConfirmationHandler = linkConfirmationHandler)
+
+        viewModel.formValuesChanged(
+            formValues = FormFieldValues(userRequestedReuse = PaymentSelection.CustomerRequestedSave.NoRequest)
+        )
+
+        viewModel.onPayClicked()
+
+        assertThat(linkConfirmationHandler.calls).hasSize(1)
+        assertThat(viewModel.state.value.primaryButtonState).isEqualTo(PrimaryButtonState.Enabled)
+        assertThat(viewModel.state.value.errorMessage).isNull()
+    }
+
+    @Test
+    fun `onPayClicked does nothing when payment selection is null`() = runTest {
```

it shouldn't be possible to click the pay button if the payment selection is null, right? bc the button will be disabled? Should we report an unexpected error here?

and/or it may be more useful to test that the button is disabled when the selection is null (if we don't already have this) than testing that nothing happens when it's called

---

`test selection and button state gets updated when form values change` covers the disabled check,  I updated this test to check for logs as well